<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="enable_rviz" default="false" />
    <group if="$(arg enable_rviz)">
        <node pkg="rviz" type="rviz" name="sloam_rviz" output="screen" args="-d $(find sloam)/launch/rviz/decentralized_sloam.rviz" />
    </group>
    <!-- robot_ns_prefix and number_of_robots is specified in sloam.yaml -->
    <arg name="param_file" default="sloam.yaml"/>
    <rosparam file="$(find sloam)/params/$(arg param_file)" command="load" />
    <arg name="hostRobot_ID" default="0" />
    <arg name="odom_topic" default="/robot0/odom"/>
    <arg name="turn_off_intra_loop_closure" default="false"/>
    <arg name="turn_off_inter_loop_closure" default="false"/>
    <arg name="turn_off_rel_inter_robot_factor" default="false"/>
    <arg name="priorTFKnown" default="false"/>
    <arg name="priorTF_x" default="0"/>
    <arg name="priorTF_y" default="0"/>
    <arg name="priorTF_z" default="0"/>
    
    <node pkg="sloam" type="sloam_node" name="sloam$(arg hostRobot_ID)" output="screen">
            <param name="hostRobotID" value="$(arg hostRobot_ID)" />
            <param name="turn_off_intra_loop_closure" value="$(arg turn_off_intra_loop_closure)"/>
            <param name="turn_off_inter_loop_closure" value="$(arg turn_off_inter_loop_closure)"/>
            <param name="turn_off_rel_inter_robot_factor" value="$(arg turn_off_rel_inter_robot_factor)"/>
            <!-- the TF here is from robot to world, which is the inverse of robot's position in the world frame -->
            <param name="priorTFKnown" value="$(arg priorTFKnown)"/>
            <param name="priorTF_x" value="$(arg priorTF_x)"/>
            <param name="priorTF_y" value="$(arg priorTF_y)"/>
            <param name="priorTF_z" value="$(arg priorTF_z)"/>
            <!-- for publish -->
            <remap from="/sloam/cubes_map"                  to="robot_$(arg hostRobot_ID)/sloam/cubes_map"/>
            <remap from="/sloam/cubes_submap"                to="robot_$(arg hostRobot_ID)/sloam/cubes_submap"/>
            <remap from="/sloam/cylinders_map"                         to="robot_$(arg hostRobot_ID)/sloam/cylinders_map"/>
            <remap from="/sloam/submap_cylinder_models"         to="robot_$(arg hostRobot_ID)/sloam/submap_cylinder_models"/>
            <remap from="/sloam/optimized_point_landmarks"         to="robot_$(arg hostRobot_ID)/sloam/optimized_point_landmarks"/>
            <remap from="/sloam/rel_inter_robot_factors" to="robot_$(arg hostRobot_ID)/sloam/rel_inter_robot_factors"/>
            <!-- for subscribe -->
            <remap from="/sloam/odom"         to="$(arg odom_topic)"/>
            <!-- semantic measurements will be under topic robot_ns_prefix+hostRobotID/semantic_meas_sync_odom -->
    </node>
    
</launch>
